src/storage/src/compaction/task.rs

```diff
@@ -12,23 +12,126 @@
+        let mut futs = Vec::with_capacity(self.outputs.len());
+        let compacted_inputs = Arc::new(Mutex::new(HashSet::new()));
+
+        for output in self.outputs.drain(..) {
+            let schema = self.schema.clone();
+            let sst_layer = self.sst_layer.clone();
+            let object_store = self.sst_layer.object_store();
+            let compacted = compacted_inputs.clone();
+            futs.push(async move {
+                match output.run(schema, sst_layer, object_store).await {
+                    Ok(meta) => {
+                        compacted
+                            .lock()
+                            .unwrap()
+                            .extend(output.inputs.iter().map(|f| FileMeta {
+                                file_name: f.file_name().to_string(),
+                                time_range: *f.time_range(),
+                                level: f.level(),
+                            }));
+                        Ok(meta)
+                    }
+                    Err(e) => Err(e),
+                }
+            });
+        }
+
+        let outputs = futures::future::join_all(futs)
+            .await
+            .into_iter()
+            .collect::<Result<_>>()?;
+        let compacted = compacted_inputs.lock().unwrap().drain().collect();
+        Ok((outputs, compacted))
+    }
+
+    /// Writes updated SST info into manifest.
+    async fn write_manifest_and_apply(
+        &self,
+        output: Vec<FileMeta>,
+        input: Vec<FileMeta>,
+    ) -> Result<()> {
+        let version = &self.shared_data.version_control;
+        // TODO(hl): do we have to update region version?
+        let region_version = version.metadata().version();
+
+        let edit = RegionEdit {
+            region_version,
+            flushed_sequence: None,
+            files_to_add: output,
+            files_to_remove: input,
+        };
+        info!(
+            "Compacted region: {}, region edit: {:?}",
+            version.metadata().name(),
+            edit
+        );
+        self.writer
+            .write_edit_and_apply(&self.wal, &self.shared_data, &self.manifest, edit, None)
+            .await
+    }
+
+    /// Mark files are under compaction.
+    fn mark_files_compacting(&self, compacting: bool) {
+        for o in &self.outputs {
+            for input in &o.inputs {
+                input.set_compacting(compacting);
+            }
+        }
+    }
 }
 
 #[async_trait::async_trait]
-impl CompactionTask for CompactionTaskImpl {
-    // TODO(hl): Actual SST compaction tasks
-    async fn run(&self) -> Result<()> {
+impl<S: LogStore> CompactionTask for CompactionTaskImpl<S> {
+    async fn run(mut self) -> Result<()> {
+        self.mark_files_compacting(true);
+        match self.merge_ssts().await {
+            Ok((output, compacted)) => {
+                self.write_manifest_and_apply(output, compacted).await?;
```

Comment:
If write_manifest_and_apply returns Err, then the files are still in compacting state. Maybe we should mark compacting to false in `Drop` and use a `success` flag to skip this. We only mark success as true when we ensure the compaction task is finished successfully.

Reply:
Maybe just match the result of `write_manifest_and_apply`? Either way we need to log the result of manifest update I think.

Reply:
Oh, you are right.